src/number_theory/number_field/embeddings.lean

```diff
@@ -323,12 +323,132 @@ begin
+(is_real_iff.mp hw).embedding
+
+@[simp]
+lemma is_real.place_embedding_apply {w : infinite_place K} (hw : is_real w) (x : K):
+  place (is_real.embedding hw) x = w x :=
+begin
+  rw [is_real.embedding, complex_embedding.is_real.place_embedding, ← coe_mk],
+  exact congr_fun (congr_arg coe_fn (mk_embedding w)) x,
+end
+
+variable (K)
+
+/-- The map from real embeddings to real infinite places -/
+noncomputable def mk_real :
+  {φ : K →+* ℂ // complex_embedding.is_real φ} → {w : infinite_place K // is_real w} :=
+subtype.map mk (λ φ hφ, ⟨φ, hφ, rfl⟩)
+
+/-- The map from nonreal embeddings to complex infinite places -/
+noncomputable def mk_complex :
+  {φ : K →+* ℂ // ¬ complex_embedding.is_real φ} → {w : infinite_place K // is_complex w} :=
+subtype.map mk (λ φ hφ, ⟨φ, hφ, rfl⟩)
+
+/-- The map from real embeddings to real infinite places as an equiv -/
+noncomputable def mk_real_equiv :
+  {φ : K →+* ℂ // complex_embedding.is_real φ} ≃ {w : infinite_place K // is_real w} :=
+{ to_fun := mk_real K,
+  inv_fun :=  λ w, ⟨w.1.embedding, is_real_iff.1 w.2⟩,
+  left_inv := λ φ, subtype.ext_iff.2 (number_field.complex_embeddings.is_real.embedding_mk φ.2),
+  right_inv := λ w, subtype.ext_iff.2 (mk_embedding w.1), }
+
+@[simp]
+lemma mk_real.apply (φ :  {φ : K →+* ℂ // complex_embedding.is_real φ}) (x : K) :
+  complex.abs (φ x) = mk_real K φ x := apply φ x
+
+@[simp]
+lemma mk_complex.apply (φ :  {φ : K →+* ℂ // ¬ complex_embedding.is_real φ}) (x : K) :
+  complex.abs (φ x) = mk_complex K φ x := apply φ x
+
+variable [number_field K]
+
+lemma mk_complex.filter (w : { w : infinite_place K // w.is_complex }) :
+  finset.univ.filter (λ φ, mk_complex K φ = w) =
+    { ⟨w.1.embedding, is_complex_iff.1 w.2⟩,
+      ⟨complex_embedding.conjugate w.1.embedding,
+        complex_embedding.is_real_conjugate_iff.not.2 (is_complex_iff.1 w.2)⟩ } :=
+begin
+  ext φ,
+  simp_rw [finset.mem_filter, subtype.val_eq_coe, finset.mem_insert, finset.mem_singleton,
+    @subtype.ext_iff_val (infinite_place K), @subtype.ext_iff_val (K →+* ℂ), @eq_comm _ φ.val,
+    ← mk_eq_iff, mk_embedding, @eq_comm _ _ w.val],
+  simpa only [finset.mem_univ, true_and],
+end
+
+lemma mk_complex.filter_card (w : { w : infinite_place K // w.is_complex }) :
+  (finset.univ.filter (λ φ, mk_complex K φ = w)).card = 2 :=
+begin
+  rw mk_complex.filter,
+  exact finset.card_doubleton
+    (subtype.mk_eq_mk.not.2 $ ne_comm.1 $
+      complex_embedding.is_real_iff.not.1 $ is_complex_iff.1 w.2),
+end
+
+noncomputable instance number_field.infinite_place.fintype : fintype (infinite_place K) :=
+set.fintype_range _
+
+noncomputable instance number_field.infinite_place.is_real.fintyp:
+  fintype { w : infinite_place K // w.is_real } := infer_instance
+
+noncomputable instance number_field.infinite_place.is_complex.fintype :
+  fintype { w : infinite_place K // w.is_complex } := infer_instance
+
+lemma prod_eq_abs_norm (x : K) :
```

Comment:
This deserves a docstring.

Reply:
I am not sure what you had in mind so I added a reference to the product formula.

Reply:
Exactly this, thanks!